src/main/scala/za/co/absa/spark/commons/implicits/DataFrameImplicits.scala

```diff
@@ -0,0 +1,143 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spark.commons.implicits
+
+import java.io.ByteArrayOutputStream
+
+import org.apache.log4j.{LogManager, Logger}
+import org.apache.spark.sql.functions.{array, callUDF, col, lit, when}
+import org.apache.spark.sql.{Column, DataFrame, SparkSession}
+import za.co.absa.spark.commons.error.ErrorMessage
+import za.co.absa.spark.commons.schema.SchemaUtils
+import za.co.absa.spark.hats.transformations.NestedArrayTransformations
+
+import scala.collection.mutable
+
+object DataFrameImplicits {
+
+  private val log: Logger = LogManager.getLogger(this.getClass)
+
+  private val overWriteErrorFunction = "overWriteErr"
+  private val overWriteErrorType = "overWriteError"
+  private val overWriteErrorCode = "E000OW"
+
+  implicit class DataFrameEnhancements(val df: DataFrame) {
+
+    private def gatherData(showFnc: () => Unit): String = {
+      val outCapture = new ByteArrayOutputStream
+      Console.withOut(outCapture) {
+        showFnc()
+      }
+      val dfData = new String(outCapture.toByteArray).replace("\r\n", "\n")
+      dfData
+    }
+
+    def dataAsString(): String = {
+      val showFnc: () => Unit = df.show
+      gatherData(showFnc)
+    }
+
+    def dataAsString(truncate: Boolean): String = {
+      val showFnc:  () => Unit = ()=>{df.show(truncate)}
+      gatherData(showFnc)
+    }
+
+    def dataAsString(numRows: Int, truncate: Boolean): String = {
+      val showFnc: ()=>Unit = () => df.show(numRows, truncate)
+      gatherData(showFnc)
+    }
+
+    def dataAsString(numRows: Int, truncate: Int): String = {
+      val showFnc: ()=>Unit = () => df.show(numRows, truncate)
+      gatherData(showFnc)
+    }
+
+    def dataAsString(numRows: Int, truncate: Int, vertical: Boolean): String = {
+      val showFnc: ()=>Unit = () => df.show(numRows, truncate, vertical)
+      gatherData(showFnc)
+    }
+
+    /**
+      * Adds a column to a dataframe if it does not exist
+      *
+      * @param colName     A column to add if it does not exist already
+      * @param colExpr     An expression for the column to add
+      * @return a new dataframe with the new column
+      */
+    def withColumnIfDoesNotExist(colName: String, colExpr: Column): DataFrame = {
+      if (df.schema.exists(field => field.name.equalsIgnoreCase(colName))) {
+        log.warn(s"Column '$colName' already exists. The content of the column will be overwritten.")
+        overwriteWithErrorColumn(df, colName, colExpr)
```

Comment:
This again is pretty Encaladus specific. I would make this branch of code and input parameter.
```scala
ifExists: (DataFrame, String) => Unit = (_, _) => {}
```
This would also eliminate the need for the specific `errorColumn` code

Reply:
So remove all the `overwriteWithErrorColumn` function?